### ISSUE TITLE
Layers shared in multiple stacks cause build/deploy to fail

### ISSUE THREAD

**Issue by kanishk-dewan** [aws/aws-sam-cli#2724]
The issue I am facing seems to be because of a recent update to SAM cli. Since my SAM templates did not change recently and my deploy was running fine up until a few days ago. Possibly a bug introduced in 1.21.x or 1.20.0.

### Description:

`sam build` errors out with following message:
`Error: L is an Invalid Layer Arn.`

_The following steps will help replicate the issue, but please refer to the "debug" output at the bottom. Because I think the underlying problem is exposed there._

I create a nested application in this case. In the parent application's template there are 2 layers defined. I pass the layer ARNs from parent to child template using Parameters. The error origins where the layer ARNs are specified inside a lambda function resource definition.

### Steps to reproduce:

Parent template:
```
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Parameters:
  AWSAccountId:
    Type: String
    Default: "123"
  IAMRoleName:
    Type: String
    Default: "role"

Resources:
  OneLayer:
    Type: AWS::Serverless::LayerVersion
    Properties:
      LayerName: "one-layer"
      ContentUri: ../layers/one/
      CompatibleRuntimes:
        - python3.8
      RetentionPolicy: Delete

  AnotherLayer:
    Type: AWS::Serverless::LayerVersion
    Properties:
      LayerName: "another-layer"
      ContentUri: ../layers/another/
      CompatibleRuntimes:
        - python3.8
      RetentionPolicy: Delete

  SomeLambdaFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: ../lambdas/example/
      FunctionName: "SomeLambdaName"
      Handler: test.lambda_handler
      Role: !Ref IAMRole
      Layers:
        - !Ref OneLayer
        - !Ref AnotherLayer

  TestApp:
    Type: AWS::Serverless::Application
    Properties:
      Location: apps/test.yaml
      Parameters:
        IAMRole: !Join ['', ["arn:aws:iam::", !Ref AWSAccountId, ":role/", !Ref IAMRoleName]]
        Layers: !Join [",", [!Ref OneLayer, !Ref AnotherLayer]]
```
Nested App template:
```
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Parameters:
  IAMRole:
    Type: String
  Layers:
    Type: CommaDelimitedList

Resources:
  NestedLambda:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: ../lambdas/example/
      FunctionName: "NestedLambdaName"
      Handler: nested.lambda_handler
      Role: !Ref IAMRole
      Layers: !Ref Layers
```

### Observed result:

```
Error: L is an Invalid Layer Arn.
```
### Expected result:

```
Build Succeeded

Built Artifacts  ...
.
.
```

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: MacOS/Ubuntu
2. `sam --version`: 1.21.1/1.21.0
3. AWS region: us-east-1

With the `--debug` flag the log statement in output that stands out is:
`Unable to resolve property Layers: {'Fn::Join': [',', [{'Ref': 'OneLayer'}, {'Ref': 'AnotherLayer'}]]}. Leaving as is.`

In my actual template I can see more statements like this for other parameters which use `Join` function in value.

**Comment by aahung** [comment]
After initial investigation, `sam` requires the property `Layers` to be a list of string. According to the [CFN doc](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-layers), `Layers` should be of type `List of String`. But apparently CFN supports comma delimited string as well. `sam build` or `sam deploy` will treat this as invalid Arn, which is a bug. This bug exists before 1.20.0, I can reproduced it in 1.19.0 without nested stacks:

```yaml
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31

Resources:
  Function:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: function
      Handler: app.lambda_handler
      Runtime: python3.8
      Layers: !Join [",", ["arn:aws:lambda:::awslayer:AmazonLinux1703", "arn:aws:lambda:::awslayer:AmazonLinux1803"]]
```

### Why this is surfaced now?

before 1.20.0, `sam build` won't go into nested stacks and `sam deploy` simply upload all resources in nested stacks to CFN without too much checks. Since 1.20.0, this behavior changed.

**Comment by kanishk-dewan** [comment]
@aahung Does `!Ref Layers` _not_ resolve to a list of strings in `sam`, since the type for parameter `Layers` is `CommaDelimitedList`.

**Comment by aahung** [comment]
> @aahung Does `!Ref Layers` _not_ resolve to a list of strings in `sam`, since the type for parameter `Layers` is `CommaDelimitedList`.

Yes this is one, also there might be other issues behind this. We are currently investigating and working on a fix. Sorry about the inconvenience.

A related question: Does your `NestedLambda` in your example require building? I wonder what was your nested stack workflow was and whether rolling back to not building nested stacks is a solution for you.

**Comment by kanishk-dewan** [comment]
Yes, in the actual workflow, the nested stack does require building, AFAIK.
For now, pinning the version to 1.19 works for me. At this point, are you able to vaguely estimate when a fix may be released?

TBH, One of the main reasons I _initially_ chose to have nested stacks is for better management of the template. The template grew to a point where it was getting difficult to manage. I wanted a way to break it up and nested stacks was the only way I could find to logically group resources into smaller, more manageable templates. Maybe there should be some feature which would allow including/importing templates. :)

**Comment by aahung** [comment]
> Yes, in the actual workflow, the nested stack does require building, AFAIK.

in 1.19.0, functions in nested stacks won't get built by `sam build`. Do you build them separately?

>  For now, pinning the version to 1.19 works for me. At this point, are you able to vaguely estimate when a fix may be released?

We are actively exploring possible solutions to this problem and treating it as a matter of top priority - we'll push out a patch release as soon as it's ready. We'll update this issue with a link to the pull request as soon as we create it.

Also I want to say your use case is very valuable for our development. Thank you for providing info about your template!

**Comment by kanishk-dewan** [comment]
Thank you for taking this up on priority.

> Do you build them separately?
No. I do a `sam build` on the parent and then a `sam deploy` on the parent too.

**Comment by aahung** [comment]
> Thank you for taking this up on priority.
> 
> > Do you build them separately?
> 
> No. I do a `sam build` on the parent and then a `sam deploy` on the parent too.

ok, if `sam build` is only executed on the parent, before 1.20.0, resources in nested stacks are not built. 

When you run `sam build`, you can see which functions are built in stdout.

```
Building codeuri: ...... functions: [<the function being built>]
```

**Comment by pathikrit-repo-personal** [comment]
Facing the same issue  also . May I know the date of release of this feature

**Comment by aahung** [comment]
@pathikrit-repo-personal it was just released in 1.22.0

**Comment by aahung** [comment]
let us know if problem still persists
